### PR TITLE
Query: Materialize Required dependents always

### LINKED ISSUE #12100: Enable configuring required 1-to-1 dependents
For normal and ownership relationships, see https://github.com/aspnet/EntityFrameworkCore/issues/9005

This would affect:

- Table splitting queries
- Potentially client-side cascade deletes
- Inserts

The API could be on the `NavigationBuilder`:
 *  `.Navigation(o => o.Details).IsRequired()`

Consider obsoleting `.IsRequired()` on the relationship
Consider warning if the constraint is not enforceable

### LINKED ISSUE #21807: EF returning NULL from queries for entities that include nested owned types
I have a EF model for an entity like:

``` C#
    public class Entity
    {
        public string Id { get; set; }
        public Contact Contact { get; set; }
    }

    public class Contact
    {
        public string Name { get; set; }
        public Address Address { get; set; }
    }

    public class Address
    {
        public string Street { get; set; }
        public string City { get; set; }
        public string State { get; set; }
        public string Zip { get; set; }
    }

    public class EntityContext : DbContext
    {
        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<Entity>(builder =>
            {
                builder.HasKey(x => x.Id);

                builder.OwnsOne(x => x.Contact, contact =>
                {
                    contact.OwnsOne(c => c.Address);
                });
            });
        }
    }
```

When using EF to query for these entities, if `Contact.Name` is null, the `Contact` property of all queried entities is always null, even if `Contact.Address` is non null.

### Steps to reproduce

``` C#
using Microsoft.Data.Sqlite;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Design;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.Threading.Tasks;

namespace EfCoreNestedOwnedTypes
{
    [TestClass]
    public class OwnedTypeTests
    {
        private static readonly string ConnectionString = "Data Source = test.db";

        private static readonly DbContextOptions<EntityContext> Options = new DbContextOptionsBuilder<EntityContext>()
            .UseSqlite(ConnectionString)
            .Options;

        [TestInitialize]
        public async Task Initialize()
        {
            using (var context = new EntityContext(Options))
            {
                await context.Database.MigrateAsync();
            }
        }

        [TestMethod]
        public async Task Querying_entity_with_value_for_nested_owned_type_should_not_return_null()
        {
            var id = Guid.NewGuid().ToString();

            using (var context = new EntityContext(Options))
            {
                context.Add(new Entity
                {
                    Id = id,
                    Contact = new Contact
                    {
                        Address = new Address
                        {
                            Zip = "12345"
                        }
                    }
                });

                await context.SaveChangesAsync();
            }

            // Get the zip using a DbCommand
            string? dbCommandZip = null;

            using (var connection = new SqliteConnection(ConnectionString))
            {
                var command = connection.CreateCommand();
                command.CommandText = @"
                    SELECT Contact_Address_Zip FROM Entity WHERE Id=$id
                ";

                command.Parameters.AddWithValue("$id", id);

                await connection.OpenAsync();
                using (var reader = command.ExecuteReader())
                {
                    while (reader.Read())
                    {
                        dbCommandZip = reader.GetString(0);
                    }
                }
            }

            // Get the zip using EF
            string? efZip = null;

            using (var context = new EntityContext(Options))
            {
                var result = await context.Set<Entity>().SingleOrDefaultAsync(x => x.Id == id);
                efZip = result?.Contact?.Address?.Zip;
            }

            Assert.AreEqual(dbCommandZip, efZip);
        }
    }

    public class Entity
    {
        public string Id { get; set; }
        public Contact Contact { get; set; }
    }

    public class Contact
    {
        public string Name { get; set; }
        public Address Address { get; set; }
    }

    public class Address
    {
        public string Street { get; set; }
        public string City { get; set; }
        public string State { get; set; }
        public string Zip { get; set; }
    }

    public class EntityContext : DbContext
    {
        public EntityContext(DbContextOptions<EntityContext> options) : base(options)
        {

        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<Entity>(builder =>
            {
                builder.HasKey(x => x.Id);

                builder.OwnsOne(x => x.Contact, contact =>
                {
                    contact.OwnsOne(c => c.Address);
                });
            });

            base.OnModelCreating(modelBuilder);
        }
    }

    public class EntityContextFactory : IDesignTimeDbContextFactory<EntityContext>
    {
        public EntityContext CreateDbContext(string[] args)
        {
            var optionsBuilder = new DbContextOptionsBuilder<EntityContext>();
            optionsBuilder.UseSqlite("Data Source=test.db");

            return new EntityContext(optionsBuilder.Options);
        }
    }
}
```

### Further technical details

EF Core version: 5.0.0-preview.7.20365.15
Database provider: I can reproduce this issue in SQLite and SqlServer
Target framework: .NET Core 3.1
Operating system: Windows
IDE: Visual Studio 2019 16.6.2

### PR DESCRIPTION
**Pull request by smitpatel**
- Also remove additional conditions in SelectExpression

Resolves #21807
Resolves #12100

### CODE REVIEW

File: src/EFCore.Relational/Metadata/Internal/RelationalModel.cs [hunk 1]
```diff
@@ -1034,6 +1034,14 @@ private static void PopulateRowInternalForeignKeys(TableBase table)
                         entityTypesToVisit.Enqueue(
                             (referencingForeignKey.DeclaringEntityType, optional || !referencingForeignKey.IsRequiredDependent));
                     }
+
+                    if (mainMapping.IncludesDerivedTypes)
+                    {
+                        foreach (var directlyDerivedEntityType in entityType.GetDirectlyDerivedTypes())
+                        {
+                            entityTypesToVisit.Enqueue((directlyDerivedEntityType, optional));
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Metadata/Internal/RelationalModel.cs:1042]: Should only visit ETs mapped to this table

File: src/EFCore.Relational/Metadata/Internal/RelationalModel.cs [hunk 2]
```diff
@@ -1034,6 +1034,14 @@ private static void PopulateRowInternalForeignKeys(TableBase table)
                         entityTypesToVisit.Enqueue(
                             (referencingForeignKey.DeclaringEntityType, optional || !referencingForeignKey.IsRequiredDependent));
                     }
+
+                    if (mainMapping.IncludesDerivedTypes)
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Metadata/Internal/RelationalModel.cs:1038]: Use the mapping for the current `entityType`, as other types sharing the table might be using different strategies

File: src/EFCore.Relational/Metadata/Internal/RelationalModel.cs (hunk 1 above)
Comment (smitpatel) [reply src/EFCore.Relational/Metadata/Internal/RelationalModel.cs:1042]: IncludesDerivedTypes was making sure that hierarchy is mapped to same table. Are both conditions needed?

Comment (AndriySvyryd) [reply src/EFCore.Relational/Metadata/Internal/RelationalModel.cs:1042]: IncludesDerivedTypes means that there are rows corresponding to the derived types in this table, it doesn't mean that they are fully contained

Comment (smitpatel) [reply src/EFCore.Relational/Metadata/Internal/RelationalModel.cs:1042]: Then do we need condition of IncludesDerivedTypes?

Comment (AndriySvyryd) [reply src/EFCore.Relational/Metadata/Internal/RelationalModel.cs:1042]: We don't *need* it, but if it's false then short-circuiting is faster

Comment (smitpatel) [reply src/EFCore.Relational/Metadata/Internal/RelationalModel.cs]: Updated both.

**Comment by Mohamad-Yousef** [comment]
Hello everyone 

One of my entities owns a value object 
 builder.OwnsOne(c => c.Amount, amount =>
            {
                amount.Property(a => a.Value).HasColumnName("Amount");
                amount.Property(a => a.CurrencyIso).HasMaxLength(3).HasColumnName("Currency").IsRequired();
            });
 builder.Navigation(x => x.Amount).IsRequired();


I have an odd behavior here 
when querying this table alone "Query#1" the Amount Navigation is populated "note: both amount and currency cols has no null values" and the following query "Query#2" that uses the same table -joined with other tables and union with other query- its Amount Navigation is populated 
BUT
if I commented out Query#1 
then in Query#2 the amount and Currency cols are excluded from the final generated raw SQL "Amount Navigation becomes NULL"

Any help would be appreciated 
side Question : Do I have to add migration and update DB if I added/removed config like 
"builder.Navigation(x => x.Amount).IsRequired();"

**Pull request merged**